### PR TITLE
SLVS-920 S2259 - Null pointers should not be dereferenced - Simple me…

### PR DESCRIPTION
**Pull request by tvsonar**
…mber access based logic

### CODE REVIEW

**Comment by tvsonar** [comment]
All of the found issues are false-positives, and most of them can be filtered by https://jira.sonarsource.com/browse/SLVS-921

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs [hunk 1]
```diff
@@ -271,7 +284,7 @@ private void VisitInstruction(Node node)
         {
             var instruction = node.ProgramPoint.Block.Instructions[node.ProgramPoint.Offset];
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs]: MAJOR 'ExplodedGraph.VisitInstruction(ExplodedGraph.Node)' has a cyclomatic complexity of 26. Rewrite or refactor the method to reduce complexity to 25. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidExcessiveComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs [hunk 2]
```diff
@@ -421,7 +461,7 @@ private bool IsLocalScoped(ISymbol symbol)
 
         private static ProgramState GetNewProgramStateForAssignment(ProgramState programState, ISymbol symbol)
         {
-            return GetNewProgramStateForAssignment(programState, symbol, null, new Optional<object>());
+            return GetNewProgramStateForAssignment(programState, symbol, null, new object());
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs]: MAJOR 'ExplodedGraph.GetNewProgramStateForAssignment(ProgramState, ISymbol)' appears to have no upstream public or protected callers. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidUncalledPrivateCode)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs [hunk 3]
```diff
@@ -421,7 +461,7 @@ private bool IsLocalScoped(ISymbol symbol)
 
         private static ProgramState GetNewProgramStateForAssignment(ProgramState programState, ISymbol symbol)
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs]: MAJOR Remove this unused private member. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category,
+                RuleSeverity.ToDiagnosticSeverity(), IsActivatedByDefault,
+                helpLinkUri: DiagnosticId.GetHelpLink(),
+                description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: Factorize all exploded graph based registrations out from here.

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                    CheckForNullDereference(declaration.Body, symbol, c);
+                },
+                SyntaxKind.AnonymousMethodExpression,
+                SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.ParenthesizedLambdaExpression);
+        }
+
+        private static void CheckForNullDereference(CSharpSyntaxNode body, ISymbol declaration, SyntaxNodeAnalysisContext context)
+        {
+            IControlFlowGraph cfg;
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: This shoudl also be factorized out from here

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs [hunk 6]
```diff
@@ -371,19 +369,61 @@ private void VisitInstruction(Node node)
...
 
                         var symbol = semanticModel.GetSymbolInfo(identifier).Symbol;
+                        newProgramState = SetNewSymbolicValueIfLocal(newProgramState, symbol);
+                    }
+                    break;
 
-                        if (IsLocalScoped(symbol))
+                case SyntaxKind.SimpleMemberAccessExpression:
+                case SyntaxKind.PointerMemberAccessExpression:
+                    {
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/CSharp/ExplodedGraph.cs]: This should most probably go into a check based class.

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            }
+
+            public override ProgramState ProcessInstruction(ProgramPoint programPoint, ProgramState programState)
+            {
+                var instruction = programPoint.Block.Instructions[programPoint.Offset];
+                switch (instruction.Kind())
+                {
+                    case SyntaxKind.IdentifierName:
+                        {
+                            // TODO handle foreach: IEnumerable<int> xs = null; foreach(var x in xs){}
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: MAJOR Remove this commented out code. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS125)
INFO Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 3 major
* INFO 1 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:39CCD086-A7F8-42A0-B402-3C9BD9EB4825: Because field 'ExplodedGraphCheck.explodedGraph' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:39CCD086-A7F8-42A0-B402-3C9BD9EB4825: Because field 'ExplodedGraphCheck.semanticModel' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:39CCD086-A7F8-42A0-B402-3C9BD9EB4825: Because field 'ExplodedGraphCheck.explodedGraph' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:39CCD086-A7F8-42A0-B402-3C9BD9EB4825: Because field 'ExplodedGraphCheck.semanticModel' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)

**Pull request merged**